### PR TITLE
fix(nat): make tagSpecification optional to fits nat-gateway without tags

### PR DESCRIPTION
**Pull request by haarchri**
Signed-off-by: haarchri <chhaar30@googlemail.com>

### Description of your changes
fixed 
` failed to create the NATGateway resource: api error InvalidParameter: Tag specification must have at least one tag`
- skip tagSpecification if tags nil - then nat-gateway is created without error message

Fixes https://github.com/crossplane/provider-aws/issues/684

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
kubectl apply -f examples/ec2/natgateway-without-tags.yaml 

```
[...]
Spec:
  Deletion Policy:  Delete
  For Provider:
    Allocation Id:  eipalloc-0386dc8235394d295
    Allocation Id Ref:
      Name:     sample-eip
    Region:     us-east-1
    Subnet Id:  subnet-0bea8bdb2151f4fbe
    Subnet Id Ref:
      Name:  sample-subnet1
  Provider Config Ref:
    Name:  example
Status:
  At Provider:
    Create Time:  2021-11-24T07:49:32Z
    Nat Gateway Addresses:
      Allocation Id:         eipalloc-0386dc8235394d295
      Network Interface Id:  eni-04b3f0dd378d5b51d
      Private Ip:            10.0.1.150
      Public Ip:             34.225.0.40
    Nat Gateway Id:          nat-0cdc37e208c2f4c36
    State:                   available
    Vpc Id:                  vpc-0557e8b3073ddddad
  Conditions:
    Last Transition Time:  2021-11-24T07:51:34Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-11-24T07:49:32Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
```

kubectl apply -f examples/ec2/natgateway.yaml 

```
[...]
Spec:
  Deletion Policy:  Delete
  For Provider:
    Allocation Id:  eipalloc-0386dc8235394d295
    Allocation Id Ref:
      Name:     sample-eip
    Region:     us-east-1
    Subnet Id:  subnet-0bea8bdb2151f4fbe
    Subnet Id Ref:
      Name:  sample-subnet1
    Tags:
      Key:    Name
      Value:  sample-natgateway
  Provider Config Ref:
    Name:  example
Status:
  At Provider:
    Create Time:  2021-11-24T09:09:38Z
    Nat Gateway Addresses:
      Allocation Id:         eipalloc-0386dc8235394d295
      Network Interface Id:  eni-07bc918765cf65ba0
      Private Ip:            10.0.1.182
      Public Ip:             34.225.0.40
    Nat Gateway Id:          nat-0194775b0544d0162
    State:                   available
    Vpc Id:                  vpc-0557e8b3073ddddad
  Conditions:
    Last Transition Time:  2021-11-24T09:11:41Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-11-24T09:09:38Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced

```

### CODE REVIEW

File: apis/ec2/v1beta1/natgateway_types.go [hunk 1]
```diff
@@ -57,8 +57,7 @@ type NATGatewayParameters struct {
 	SubnetIDSelector *xpv1.Selector `json:"subnetIdSelector,omitempty"`
 
 	// Tags represents to current ec2 tags.
-	// +optional
-	Tags []Tag `json:"tags,omitempty"`
+	Tags []Tag `json:"tags"`
```
Comment (negz) [inline apis/ec2/v1beta1/natgateway_types.go:60]: Nit: Should we update the description to explicitly mention that at least one tag appear to be required by the API.

**Review by negz — COMMENTED** [review]
> other option could be that we inject a name-tag during creation - what you guys think ?

I think I prefer requiring tags, presuming what the API actually wants is "at least one tag, regardless of its key and value". Is that the case, or does the API require a tag with key `name`?

How sure are we that tags are always required? I don't have a ton of faith in the AWS docs, but I do note that they say tags are not required and the given example includes no tags. 🤔 Are they perhaps only required under certain circumstances?

https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateNatGateway.html

**Comment by negz** [comment]
P.S. This is a 'breaking change' at the API level because it makes a previously optional field required, but I'm okay with that if the optional field (tags) was actually required in practice - i.e. if omitting tags results in a broken NAT gateway 100% of the time.

**Comment by haarchri** [comment]
@negz  i checked again the whole code and i found the problem is that we use tagSpecification which needs one tag - if we skip it if tags are not set then the process is working

File: pkg/controller/ec2/natgateway/controller.go [hunk 2]
```diff
@@ -135,17 +135,28 @@ func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.Ex
...
-	})
+		}
+	} else {
+		input = &awsec2.CreateNatGatewayInput{
+			ConnectivityType: awsec2types.ConnectivityType(cr.Spec.ForProvider.ConnectivityType),
+			AllocationId:     cr.Spec.ForProvider.AllocationID,
+			SubnetId:         cr.Spec.ForProvider.SubnetID,
+		}
+	}
+
```
Comment (negz) [inline pkg/controller/ec2/natgateway/controller.go:158]: Style nit: I would prefer:

```go
// Create an input without tags.
input := &awsec2.CreateNatGatewayInput{
	ConnectivityType: awsec2types.ConnectivityType(cr.Spec.ForProvider.ConnectivityType),
	AllocationId:     cr.Spec.ForProvider.AllocationID,
	SubnetId:         cr.Spec.ForProvider.SubnetID,
}

// If we specified tags, update the above input.
if cr.Spec.ForProvider.Tags != nil {
	input.TagSpecifications = []awsec2types.TagSpecification{{
		ResourceType: "natgateway",
		Tags:         v1beta1.GenerateEC2Tags(cr.Spec.ForProvider.Tags),
	}}
}
```

At a glance it seems like this should work.

Comment (haarchri) [reply pkg/controller/ec2/natgateway/controller.go:152]: thanks for review - tested again and changed the code - works like expected

**Review by negz — APPROVED** [review]
Thanks @haarchri!

**Pull request merged**